packages-common/pluggable-widgets-generator/generators/app/index.js

```diff
@@ -0,0 +1,490 @@
+const path = require("path");
+const pkg = require(path.join(__dirname, "/../../package.json"));
+const fs = require("fs").promises;
+const Generator = require("yeoman-generator");
+
+const promptTexts = require("./lib/prompttexts.js");
+const text = require("./lib/text.js");
+
+const widgetSrcFolder = "src/components/";
+const templates = {
+    webEmptyJs: "pluggable/web/emptyTemplateJs/",
+    webFullJs: "pluggable/web/fullTemplateJs/",
+    webEmptyTs: "pluggable/web/emptyTemplateTs/",
+    webFullTs: "pluggable/web/fullTemplateTs/",
+    nativeEmptyJs: "pluggable/native/emptyTemplateJs/",
+    nativeFullJs: "pluggable/native/fullTemplateJs/",
+    nativeEmptyTs: "pluggable/native/emptyTemplateTs/",
+    nativeFullTs: "pluggable/native/fullTemplateTs/"
+};
+
+const banner = text.getBanner(pkg);
+
+class MxGenerator extends Generator {
```

Comment:
Refactored it, reduced the code quite a bit.

Reply:
Its great improvements! I would try to cut this file even further. Maybe try some of the following tricks, though no guarantee that they work 😉 :
1. Convert some of its methods to stand-alone functions and move them to a separate file. I think that there are definitely some methods without a dependency on `this` (or where that dependency can be cut by replacing yeoman `this` helpers with normal fs/spawn). Maybe if there are enough of them and if they form a logical clique, they can be split off.
2. I see that in some places we do a regexp replace, while in others we use `ejs`. Templates feel a better fit for templating, and since we have templating engine already its a shame not to use it.
3. `_defineProperties` looks suspicious - it creates a huge `this.widget` from `this.props`, while there is not a very significant difference between the two. Maybe we can get rid of `this.widget`.
4. `templates` seem like they can be simplified to a single templated string.

Reply:
1. I moved some functions to `utils.js`. The `_copyFile` and `_copyTemplate` functions really need to be inside the class for the this binding. Moving the functions to the `utils.js` file isn't nice, since you will need to explicitly bind the this for each function call.
2. Everything is generated now with the templating engine.
3. I removed `_defineProperties`. I still make use of two objects: a `widget` object that has the answers object as it's prototype (see `utils.js`).
4. Refactored this into a single templated string.